core-backend/src/funcs.rs

```diff
@@ -24,67 +24,309 @@ use crate::{
+    fn from((): ()) -> Self {
+        Self(ReturnValue::Unit)
+    }
+}
+
+impl From<i32> for SysCallReturnValue {
+    fn from(value: i32) -> Self {
+        Self(ReturnValue::Value(Value::I32(value)))
+    }
+}
+
+impl From<u32> for SysCallReturnValue {
+    fn from(value: u32) -> Self {
+        Self(ReturnValue::Value(Value::I32(value as i32)))
+    }
+}
+
+pub(crate) trait SysCallContext: Sized {
+    fn from_args(args: &[Value]) -> Result<(Self, &[Value]), HostError>;
+}
+
+pub(crate) trait SysCall<Ext, T = ()> {
+    type Context: SysCallContext;
+
+    fn execute(
+        self,
+        caller: &mut CallerWrap<Ext>,
+        ctx: Self::Context,
+    ) -> Result<(u64, T), HostError>;
+}
+
+pub(crate) trait SysCallBuilder<Ext, Args: ?Sized, R, S> {
+    fn build(self, args: &[Value]) -> Result<S, HostError>;
+}
+
+impl<Ext, R, S, B> SysCallBuilder<Ext, (), R, S> for B
+where
+    B: FnOnce() -> S,
+    S: SysCall<Ext, R>,
+{
+    fn build(self, args: &[Value]) -> Result<S, HostError> {
+        let _: [Value; 0] = args.try_into().map_err(|_| HostError)?;
+        Ok((self)())
+    }
+}
+
+impl<Ext, R, S, B> SysCallBuilder<Ext, [Value], R, S> for B
+where
+    B: for<'a> FnOnce(&'a [Value]) -> S,
+    S: SysCall<Ext, R>,
+{
+    fn build(self, args: &[Value]) -> Result<S, HostError> {
+        Ok((self)(args))
+    }
+}
+
+// implement [`SysCallBuilder`] for functions with different amount of arguments
+macro_rules! impl_syscall_builder {
+    ($($generic:ident),+) => {
+        #[allow(non_snake_case)]
+        impl<Ext, Res, Call, Builder, $($generic),+> SysCallBuilder<Ext, ($($generic,)+), Res, Call>
+            for Builder
+        where
+            Builder: FnOnce($($generic),+) -> Call,
+            Call: SysCall<Ext, Res>,
+            $( $generic: TryFrom<SysCallValue, Error = HostError>,)+
         {
-            ::log::trace!(target: "syscalls", "{}", $name);
+            fn build(self, args: &[Value]) -> Result<Call, HostError> {
+                const ARGS_AMOUNT: usize = impl_syscall_builder!(@count $($generic),+);
+
+                let [$($generic),+]: [Value; ARGS_AMOUNT] = args.try_into().map_err(|_| HostError)?;
+                $(
+                    let $generic = SysCallValue($generic).try_into()?;
+                )+
+                Ok((self)($($generic),+))
+            }
         }
+    };
+    (@count $generic:ident) => { 1 };
+    (@count $generic:ident, $($generics:ident),+) => { 1 + impl_syscall_builder!(@count $($generics),+) };
+}
+
+impl_syscall_builder!(A);
+impl_syscall_builder!(A, B);
+impl_syscall_builder!(A, B, C);
+impl_syscall_builder!(A, B, C, D);
+impl_syscall_builder!(A, B, C, D, E);
+impl_syscall_builder!(A, B, C, D, E, F);
+impl_syscall_builder!(A, B, C, D, E, F, G);
+
+type SimpleSysCall<F> = F;
+
+impl<T, F, Ext> SysCall<Ext, T> for SimpleSysCall<F>
+where
+    F: FnOnce(&mut CallerWrap<Ext>) -> Result<T, HostError>,
+    Ext: BackendExternalities + 'static,
```

Comment:
please normalise order of impls for fallible/infallible and add comment sections